File: examples/java/mysql/src/main/java/ftl/mysql/Mysql.java
package ftl.mysql;

import java.util.List;

import xyz.block.ftl.FunctionVerb;
import xyz.block.ftl.Verb;

public class Mysql {

    @Verb
    public static class Insert implements FunctionVerb<InsertRequest, InsertResponse> {

        final CreateRequestClient c;

        public Insert(CreateRequestClient c) {
            this.c = c;
        }

        @Override
        public InsertResponse call(InsertRequest req) {
            return new InsertResponse();
        }
    }

    @Verb
    public static class Query implements FunctionVerb<GetRequestDataClient, List<String>> {

        final GetRequestDataClient c;

        public Query(GetRequestDataClient c) {
            this.c = c;
        }

        @Override
        public List<String> call(GetRequestDataClient req) {
            return req.call();
        }
    }

}

File: examples/kotlin/mysql/src/main/kotlin/ftl/mysql/Mysql.kt
package ftl.mysql

import xyz.block.ftl.*

data class InsertRequest(val `data`: String, val id: Long)

data class QueryResult(val `data`: String, val id: Long)

data class IDsQuery(val ids: List<Long>)

@Export @Verb
fun insert(req: InsertRequest, c: CreateRequestClient) {
    c.call(req.data)
}

@Export @Verb
fun query(c : GetRequestDataClient): List<QueryResult> {
    return c.call().map { r -> QueryResult(r.data, r.id) }
}

@Export @Verb
fun queryByIDs(req : IDsQuery, c: GetRequestsWithIDsClient): List<QueryResult> {
  return c.call(req.ids).map { r ->  QueryResult(r.data, r.id) }
}
